language/move-vm/runtime/src/interpreter.rs

```diff
@@ -330,6 +307,19 @@ impl Interpreter {
         let return_values = result
             .result
             .map_err(|code| PartialVMError::new(StatusCode::ABORTED).with_sub_status(code))?;
+        // Paranoid check to protect us against incorrect native function implementations. A native function that
+        // returns a different number of values than its declared types will trigger this check
+        if return_values.len() != return_type_count {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR).with_message(
```

Comment:
Those internal invariant violation errors of the VM are problematic because they do not lead to `Keep`, and thus no gas charge and potential source of DoS. Since this is kind of the same category as a verifier error (only later and dynamic) perhaps NUMBER_OF_ARGUMENTS_MISMATCH or CONSTRAINT_VIOLATION which leads to `Keep`.

Reply:
> Those internal invariant violation errors of the VM are problematic because they do not lead to Keep, and thus no gas charge and potential source of DoS.

I know that VMStatus is in the Move repo, but it is adapter specific. We do not use it on the SUI side. 
Nonetheless, we would likely also not charge for this invariant violation ;) 
I'm not sure what else to do here though? It is not clear if you hit this if it is an attack or a VM bug.

Reply:
Only the framework can define `native function`, so I think it is impossible to use this Err for DoS.

Reply:
I think the attack Wolfgang is concerned about is that if there's a bug in a framework native function that triggers this, that is a potential DOS vector.

Reply:
Got it. But if we `Keep` the transaction that triggers this ERR, will make it more difficult for us to fix the bugs. We need to ensure that the transactions are executed on the bug-fixed VM with the same result.